cpp/test/phonenumbers/matcher_test.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright (C) 2017 The Libphonenumber Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Tests that all implementations of MatcherApi are consistent.
+
+#include "phonenumbers/matcher_api.h"
+
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "phonenumbers/regex_based_matcher.h"
+#include "phonenumbers/phonemetadata.pb.h"
+
+namespace i18n {
+namespace phonenumbers {
+
+class MatcherTest : public testing::Test {
+ protected:
+  void CheckMatcherBehavesAsExpected(const MatcherApi& matcher) const {
+    PhoneNumberDesc desc;
+
+    desc = CreateDesc("");
+    // Test if there is no matcher data.
+    ExpectInvalid(matcher, "1", desc);
+
+    // Matcher data generated by MatcherCompiler.
+    desc = CreateDesc("9\\d{2}");
+    ExpectInvalid(matcher, "91", desc);
+    ExpectInvalid(matcher, "81", desc);
+    ExpectMatched(matcher, "911", desc);
+    ExpectInvalid(matcher, "811", desc);
+    ExpectTooLong(matcher, "9111", desc);
+    ExpectInvalid(matcher, "8111", desc);
+
+    desc = CreateDesc("\\d{1,2}");
+    ExpectMatched(matcher, "2", desc);
+    ExpectMatched(matcher, "20", desc);
+
+    desc = CreateDesc("20?");
+    ExpectMatched(matcher, "2", desc);
+    ExpectMatched(matcher, "20", desc);
+
+    desc = CreateDesc("2|20");
+    ExpectMatched(matcher, "2", desc);
+    // Subtle case where lookingAt() and matches() result in different end()s.
+    ExpectMatched(matcher, "20", desc);
+  }
+
+ private:
+  // Helper method to set national number fields in the PhoneNumberDesc proto.
+  // Empty fields won't be set.
+  PhoneNumberDesc CreateDesc(
+      const string& national_number_pattern) const {
+    PhoneNumberDesc desc;
+    if (!national_number_pattern.empty()) {
+      desc.set_national_number_pattern(national_number_pattern);
+    }
+    return desc;
+  }
+
+  void ExpectMatched(
```

Comment:
None of these methods need to be on the class? they can be just in the anonymous namespace above.